esmvalcore/preprocessor/_supplementary_vars.py

```diff
@@ -103,6 +104,123 @@ def add_cell_measure(
 
 
+def get_data_dims(  # noqa: C901
+    cube_c: Cube,
+    cube_a: Cube | iris.coords.AncillaryVariable,
+    cube_a_var: iris.coords.AncillaryVariable,
+) -> list[None | int]:
+    """Get matching data dimensions between cube and ancillary variable.
+
+    Parameters
+    ----------
+    cube_c: Cube
+        Iris cube with input data.
+    cube_a: Cube or iris.coords.AncillaryVariable
+        Iris cube or AncillaryVariable with ancillary data.
+    cube_a_var: iris.coords.AncillaryVariable
+        Ancillary variable object containing the ancillary data.
+
+    Returns
+    -------
+    list
+        List with as many entries as the ancillary variable dimensions.
+        The i-th entry corresponds to the i-th ancillary variable
+        dimension match with the cube's dimensions. If there is no match
+        between the ancillary variable and the cube for a dimension, then entry
+        defaults to None.
+    """
+    # Match the coordinates of the ancillary cube to coordinates and
+    # dimensions in the input cube before adding the ancillary variable.
+    data_dims: list[None | int] = []
+    if isinstance(cube_a, iris.coords.AncillaryVariable):
+        start_dim = cube_c.ndim - len(cube_a_var.shape)
+        data_dims = list(range(start_dim, cube_c.ndim))
+    else:
+        data_dims = [None] * cube_a.ndim
+        for coord in cube_a.coords():
+            logger.debug("Matching coordinate...\n%s", coord)
+            try:
+                for ancillary_dim, cube_dim in zip(
+                    cube_a.coord_dims(coord),
+                    cube_c.coord_dims(coord),
+                    strict=False,
+                ):
+                    data_dims[ancillary_dim] = cube_dim
+                logger.debug("Matched coordinate.")
+            except iris.exceptions.CoordinateNotFoundError:
+                logger.debug(
+                    "No metadata match from ancillary coord in cube coords.",
+                )
+                cube_dims = []
+                for cube_coord in cube_c.coords():
+                    if (
+                        np.issubdtype(cube_coord.dtype, np.number)
+                        and (
+                            cube_coord.standard_name == coord.standard_name
+                            or cube_coord.var_name == coord.var_name
+                        )
+                        and cube_coord.units == coord.units
+                        and cube_coord.points.shape == coord.points.shape
+                    ):
+                        # Trying to cast back and forth the coordinate points
+                        logger.debug("Trying to cast coordinates...")
+                        coord_ = coord.copy()
+                        coord_.points = (
+                            coord_.core_points()
+                            .astype(np.float32)
+                            .astype(np.float64)
+                        )
+                        if np.all(coord_.points == cube_coord.points):
```

Comment:
```suggestion
                        )
                        cube_coord = cube_coord.copy()
                        if np.all(coord_.points == cube_coord.points):
```
To avoid modifying the input cube by realizing its coordinate data on the next line